### PR TITLE
feat: Custom Login Page feature

### PR DESCRIPTION
**Pull request by Pucek9**
Closes [CXSPA-9863](https://jira.tools.sap/browse/CXSPA-9863)

### CODE REVIEW

File: feature-libs/user/account/components/login-form/custom-login-form-component.service.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  UntypedFormControl,
+  UntypedFormGroup,
+  Validators,
+} from '@angular/forms';
+import { CustomFormValidators } from '@spartacus/storefront';
+
+@Injectable()
+export class CustomLoginFormComponentService extends LoginFormComponentService {
+  config = inject(Config);
+  authConfigService = inject(AuthConfigService);
```
Comment (ercultimate) [inline feature-libs/user/account/components/login-form/custom-login-form-component.service.ts]: Please set the appropriate visibility on the properties.

File: feature-libs/user/account/components/login-form/custom-login-form-component.service.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+  method = 'POST';
+  action = this.authConfigService?.getCustomLoginFormEndpoint();
+
+  form: UntypedFormGroup = new UntypedFormGroup({
```
Comment (ercultimate) [inline feature-libs/user/account/components/login-form/custom-login-form-component.service.ts]: Please use Typed forms in new components.

File: feature-libs/user/account/components/login-form/login-form.module.ts [hunk 3]
```diff
@@ -58,6 +59,17 @@ import { LoginFormComponent } from './login-form.component';
             },
           ],
         },
+        OAuthLoginComponent: {
```
Comment (ercultimate) [inline feature-libs/user/account/components/login-form/login-form.module.ts]: We should change the component type now to something more clear.  "CustomLoginComponent", maybe?

File: projects/core/src/auth/user-auth/config/auth-config.ts [hunk 4]
```diff
@@ -90,6 +90,13 @@ export abstract class AuthConfig {
      * Config for angular-oauth-oidc library.
      */
     OAuthLibConfig?: AuthLibConfig;
+
+    customLoginPage?: {
+      /** auth server's csrf endpoint */
```
Comment (ercultimate) [inline projects/core/src/auth/user-auth/config/auth-config.ts]: Please clean up my rough comments on the new properties.

File: projects/core/src/auth/user-auth/config/default-auth-config.ts [hunk 5]
```diff
@@ -42,7 +42,7 @@ export function provideAuthorizationCodeFlowByDefault(
 
 export const defaultAuthConfig: AuthConfig = {
   authentication: {
-    client_id: 'mobile_android_public',
+    client_id: 'customloginpage_public',
```
Comment (ercultimate) [inline projects/core/src/auth/user-auth/config/default-auth-config.ts]: I think we should keep `mobile_android_public` as the default, and instead update the sample data.

This is because many other teams are expecting custom login page feature to be enabled for their user flow, and we are relying on it for access to the register page and forgot password page.

File: projects/core/src/auth/user-auth/facade/auth.service.ts [hunk 6]
```diff
@@ -109,6 +111,25 @@ export class AuthService {
...
+  /**
+   * Retrieves the CSRF (Cross-Site Request Forgery) token from the service.
+   *
+   * @return {string} The CSRF token used for preventing cross-site request forgery attacks.
+   */
+  getCsrfToken() {
+    return this.crossSiteRequestForgeryService.getCsrfToken().pipe(
+      tap({
+        error: (e) => {
+          console.log('Failed to get csrf token', e);
```
Comment (ercultimate) [inline projects/core/src/auth/user-auth/facade/auth.service.ts]: Is console use intended here?

File: projects/core/src/auth/user-auth/facade/auth.service.ts [hunk 7]
```diff
@@ -109,6 +111,25 @@ export class AuthService {
...
+          if (e.status === 403) {
+            /* Redirect to restart the flow if an attempt was made to manually obtain a custom form */
+            this.routingService.go({ cxRoute: 'login' });
```
Comment (ercultimate) [inline projects/core/src/auth/user-auth/facade/auth.service.ts:128]: I'm not sure what to make of the error handling.  Have you seen many errors where restarting the flow would fix it?  If not, it might be best to show an error on the form saying something like "and error occurred,  try again later".

File: projects/core/src/auth/user-auth/services/oauth-lib-wrapper.service.ts [hunk 8]
```diff
@@ -35,7 +35,7 @@ export class OAuthLibWrapperService {
 
   protected initialize() {
     const isSSR = !this.winRef.isBrowser();
-    this.oAuthService.configure({
+    const config = {
```
Comment (ercultimate) [inline projects/core/src/auth/user-auth/services/oauth-lib-wrapper.service.ts]: Please remove my old debugging code ;)

File: projects/storefrontapp/src/app/spartacus/spartacus-b2c-configuration.module.ts [hunk 9]
```diff
@@ -33,6 +34,11 @@ const baseSite = environment.epdVisualization
   ? ['electronics-epdvisualization-spa'].concat(defaultBaseSite)
   : defaultBaseSite;
 
+console.log('zxxxx', environment, environment.newAuthFlow);
```
Comment (ercultimate) [inline projects/storefrontapp/src/app/spartacus/spartacus-b2c-configuration.module.ts]: Please remove console use.

File: projects/storefrontapp/src/app/spartacus/spartacus-b2c-configuration.module.ts [hunk 10]
```diff
@@ -33,6 +34,11 @@ const baseSite = environment.epdVisualization
...
+const newAuthFlowProviders = environment.newAuthFlow
+  ? [provideAuthorizationCodeFlowByDefault()]
```
Comment (ercultimate) [inline projects/storefrontapp/src/app/spartacus/spartacus-b2c-configuration.module.ts]: `provideAuthorizationCodeFlowByDefault()` will be converted to feature flags.  Do we need to or want to use an environment variable to enable or disable this?

File: projects/storefrontlib/cms-structure/routing/default-routing-config.ts [hunk 11]
```diff
@@ -23,6 +23,11 @@ export const defaultRoutesConfigFactory: () => RoutingConfig = () => {
           protected: false,
           authFlow: true,
         },
+        customLogin: {
+          paths: ['custom-login'],
```
Comment (ercultimate) [inline projects/storefrontlib/cms-structure/routing/default-routing-config.ts]: We should update the URL path segment to a more appropriate customer-facing value.  'login-form' is my suggestion.

File: projects/core/src/auth/user-auth/facade/auth.service.ts (hunk 7 above)
Comment (ercultimate) [reply projects/core/src/auth/user-auth/facade/auth.service.ts:128]: Additionally, we should probably let the customer know why clicking the login form just flashed the screen and cleared the form.

File: projects/core/src/auth/user-auth/facade/auth.service.ts [hunk 12]
```diff
@@ -109,6 +111,25 @@ export class AuthService {
     } catch {}
   }
 
+  /**
+   * Retrieves the CSRF (Cross-Site Request Forgery) token from the service.
```
Comment (ercultimate) [inline projects/core/src/auth/user-auth/facade/auth.service.ts]:
```suggestion
   * Retrieves the CSRF (Cross-Site Request Forgery) token for the custom login form.
```

**Review by ercultimate — COMMENTED** [review]
We'll also need some unit tests and eventually E2E test updates.

File: projects/core/src/auth/user-auth/config/auth-config.ts (hunk 4 above)
Comment (Pucek9) [reply projects/core/src/auth/user-auth/config/auth-config.ts]: In this file, it is standard to describe properties

Comment (ercultimate) [reply projects/core/src/auth/user-auth/config/auth-config.ts]: Yes, but these comments are inconsistent with the other doc-comments and need cleaned up.

File: projects/storefrontapp/src/app/spartacus/spartacus-b2c-configuration.module.ts [hunk 13]
```diff
@@ -57,6 +58,8 @@ const baseSite = environment.epdVisualization
         },
       },
     }),
+    // this will be under featureFlag (e2e branch)
```
Comment (ercultimate) [inline projects/storefrontapp/src/app/spartacus/spartacus-b2c-configuration.module.ts:61]: Are the comments intended?

File: .env-cmdrc [hunk 14]
```diff
@@ -101,5 +101,8 @@
   "punchout": {
     "CX_PUNCHOUT": "true",
     "CX_B2B": "true"
+  },
+  "newAuthFlow": {
+    "CX_NEW_AUTH_FLOW": "true"
```
Comment (ercultimate) [inline .env-cmdrc]: This environment variable is unused.  What is the intention for it?

File: projects/storefrontapp/src/app/spartacus/spartacus-b2c-configuration.module.ts (hunk 13 above)
Comment (Pucek9) [reply projects/storefrontapp/src/app/spartacus/spartacus-b2c-configuration.module.ts:61]: yes, it will be cleared with merging conflicts

File: .env-cmdrc (hunk 14 above)
Comment (Pucek9) [reply .env-cmdrc]: sorry missed that one

File: projects/core/src/auth/user-auth/facade/auth.service.ts (hunk 6 above)
Comment (Pucek9) [reply projects/core/src/auth/user-auth/facade/auth.service.ts]: yes, I'd like to debug it later after deployment

File: projects/core/src/auth/user-auth/facade/auth.service.ts (hunk 7 above)
Comment (Pucek9) [reply projects/core/src/auth/user-auth/facade/auth.service.ts:128]: This is a very edge case, when the user started his session without a previous redirection. 
I was not able to test it well locally, as due to CORS, the response seen by Angular was always 0, even I had seen 403 or other errors in the browser's dev tools.

Played with angular proxy, but it seems that for our builder it requires 3rd part stuff

I don't want to do that redirection in case, that eg, `csrf` url is configured wrongly, as it can cause an infinite loop redirection

Comment (ercultimate) [reply projects/core/src/auth/user-auth/facade/auth.service.ts:128]: Ah, I see.  Then this is reasonable.

File: projects/core/src/auth/user-auth/facade/auth.service.ts (hunk 6 above)
Comment (ercultimate) [reply projects/core/src/auth/user-auth/facade/auth.service.ts]: If this is intended for production, then use `console.error`, otherwise please ensure it is removed before feature freeze.

**Comment by Pucek9** [comment]
> We'll also need some unit tests and eventually E2E test updates.

Yeah this is a draft PR, after finishing unit tests, I'll mark it as ready

File: package.json [hunk 15]
```diff
@@ -143,6 +143,7 @@
     "start:punchout": "env-cmd --no-override -e punchout,$SPA_ENV nx serve storefrontapp --configuration=development",
     "start:prod": "env-cmd --no-override -e dev,b2c,$SPA_ENV nx serve storefrontapp --configuration=production",
     "start:pwa": " cd ./dist/storefrontapp/browser && http-server -p 4200 --silent --proxy http://localhost:4200?",
+    "start:new-auth-flow": "env-cmd --no-override -e s1,newAuthFlow,$SPA_ENV nx serve storefrontapp --configuration=development",
```
Comment (ercultimate) [inline package.json]: There are several references to `newAuthFlow` throughout the PR.  They should be removed, correct?

Comment (Pucek9) [reply package.json]: yeah unfortunately, after reverting changes and merging, it came back😅

**Comment by Pucek9** [comment]
Changed target to epic, will be continued on https://github.com/SAP/spartacus/pull/20565
